### PR TITLE
 #168781697 - Add Search Functionality on Requests

### PR DESCRIPTION
**Pull request by nignanthomas**
#### What does this PR do?
Add Search Functionality

#### Description of Task to be completed?
- Create tests
- Create searchServices
- Create validation for the search options
- Create controller

#### How should this be manually tested?
- Clone the repository from [here](https://github.com/andela/caret-bn-backend.git)
- Browse to the repo directory
- navigate to ft-approve-reject-request-168781691 branch
- Run `npm install`
- run `sequelize db:migrate`
- run `sequelize db:seed:all`
- run `npm run dev`

**To Search Requests**
* If you are logged in as a manager
Send a GET request to `http://localhost:port/api/v1/requests/manager/search?id=2&reasons=meeting`
* If you are logged in as a requester
Send a GET request to `http://localhost:port/api/v1/requests/search?id=2&reasons=meeting`

The available search quesries are:
```
    id: integer(a request id),
    userId: integer(a user id - /manager/search route only),
    origin: integer (a location id - kindly query the locations routes to have a list of locations)
    statusId: integer (a status id - 1 for Pending, 2 for Rejected, 3 for Approved )
    reasons: "string"(a keyword to search for in the reasons)
    departureDate: "YYYY-MM-DD"
    destination: integer (a location id - kindly query the locations routes to have a list of locations)
```

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#168781697](https://www.pivotaltracker.com/story/show/168781697)

#### Screenshots (if appropriate)
Screen Shot 2019-10-26 at 15 36 50
Screen Shot 2019-10-26 at 15 38 02
Screen Shot 2019-10-26 at 15 38 15
Screen Shot 2019-10-26 at 15 39 48

#### Questions:
N/A

### CODE REVIEW

**Review by sabin18 — CHANGES_REQUESTED** [review]
good job @nignanthomas but i found some changes that you can work on validation messages should be specific, reason letters should be above 3 letters .DepartureDate  search is  not working and change the PR where the startDate should be departureDate.

**Review by ngireric123 — APPROVED** [review]
good job @nignanthomas

File: src/controllers/requestController.js [hunk 1]
```diff
@@ -23,4 +26,25 @@ export default class requestController {
       200
     );
   }
+
```
Comment (nakiwuge) [inline src/controllers/requestController.js:80]: Remove this line

File: src/middlewares/managerUserIdField.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import models from '../database/models';
+import Utilities from '../utils/index';
+
+const managerUserIdField = async (req, res, next) => {
+  // eslint-disable-next-line max-len
+  if ((Object.keys(req.body).length === 0) || (Object.keys(req.body).length > 0 && !Object.keys(req.body).includes('userId'))) {
+    const managerId = req.user.payload.id;
+    const myUsers = [];
```
Comment (nakiwuge) [inline src/middlewares/managerUserIdField.js:8]: Add a new line

File: src/middlewares/managerUserIdField.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    await models.users.findAll({
+      where: {
+        lineManager: managerId,
+      }
+    }).then(users => users.forEach(user => myUsers.push(user.id)));
+    req.body.userId = myUsers;
+    return next();
+  }
```
Comment (nakiwuge) [inline src/middlewares/managerUserIdField.js:17]: Add a new line

File: src/middlewares/managerUserIdField.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  const user = await models.users.findOne({
+    where: {
+      id: req.body.userId,
+    }
+  });
```
Comment (nakiwuge) [inline src/middlewares/managerUserIdField.js:23]: Add a new line

File: src/routes/api/requests.js [hunk 5]
```diff
@@ -1,9 +1,101 @@
...
 import validateToken from '../../middlewares/auth/validateToken';
+import InputValidation from '../../middlewares/inputValidation';
+import checkUserIdField from '../../middlewares/checkUserIdField';
+import checkRole from '../../middlewares/checkRole';
+import managerUserIdField from '../../middlewares/managerUserIdField';
 
 const router = new Router();
+const { validateSearchRequestUser, validateSearchRequestManager } = InputValidation;
+const { checkManagerRole, supplierNotAllowed } = checkRole;
+
```
Comment (nakiwuge) [inline src/routes/api/requests.js:18]: remove this line

File: src/routes/api/requests.js [hunk 6]
```diff
@@ -1,9 +1,101 @@
...
+ *       '200':
+ *         description: Requests Fetched successfully!
+ *       '404':
+ *         description: No Requests Registered!
+*/
 
 router.get('/', validateToken, (req, res) => requestController.viewRequests(req, res));
+router.post('/search', validateToken, supplierNotAllowed, validateSearchRequestUser, checkUserIdField, requestController.searchRequests);
+// eslint-disable-next-line max-len
+router.post('/manager/search', validateToken, supplierNotAllowed, checkManagerRole, validateSearchRequestManager, managerUserIdField, requestController.searchRequests);
```
Comment (nakiwuge) [inline src/routes/api/requests.js]: This should be a `get` method. Use query params for search

File: src/tests/searchRequestsTests.spec.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+chai.should();
+chai.use(chaiHttp);
+
+let userToken;
+let supplierToken;
+let managerToken;
+let anotherManagerToken;
+
+
```
Comment (nakiwuge) [inline src/tests/searchRequestsTests.spec.js]: remove this line

File: src/tests/searchRequestsTests.spec.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      });
+    chai.request(app)
+      .post('/api/v1/users/login')
+      .send(mockData.anotherManager)
+      .end((err, res) => {
+        anotherManagerToken = res.body.data.token;
+        done();
+      });
+  });
+
```
Comment (nakiwuge) [inline src/tests/searchRequestsTests.spec.js:43]: remove this line

File: src/utils/requestsSearch.js [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    const { duration } = options;
+    durations = duration;
+  }
+
+  const requestsByDuration = [];
+
+  if (durations) {
+    requests.forEach(request => {
+      const { departureDate } = request.dataValues;
+      let { returnDate } = request.dataValues;
```
Comment (nakiwuge) [inline src/utils/requestsSearch.js:106]: Add a new line

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@nignanthomas great job sofar however kindly use query params for search. I have left other comments on the pr.

**Review by sabin18 — APPROVED** [review]
good job @nignanthomas.

File: src/database/models/destinations.js [hunk 10]
```diff
@@ -1,9 +1,10 @@
+
 module.exports = (sequelize, DataTypes) => {
   const destinations = sequelize.define('destinations', {
     locationId: DataTypes.INTEGER,
-    accomodationId: DataTypes.INTEGER,
```
Comment (zellymiriam) [inline src/database/models/destinations.js]: Why have you removed the `accomodationId`?

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**